### PR TITLE
B2g forceIssuer support

### PR DESCRIPTION
**Pull request by ozten**
What is it?

if you

1) go to http://192.168.186.138:10001/
2) check Force Issuer to issuer.domain and click "Get an Assertion"
3) enter a primary email like 'test@example.domain' and click Next
4) Create a password
5) Get VERIFICATION URL out of logs and go there

Note: You're logged in and the assertion was backed by 'issuer.domain'.

Note: The assertion is verified by the verifier.

This is because config/local.json adds this setting.

Caveats:

This patch is rushed and the frontend code is fairly horrible. Lots of duplicate code, I don't like the use of the `User.forceIssuer` as a global variable, etc.

How did storage get so ugly?
Why forceIssuerEmails?  
localStorage is a flat key value store, so we couldn't do:
- storage.default.emails and storage.fxos.emails
- dynamic keys (such as using fxos.login.persona.org) don't work well with removeItem and other APIs here.
- We can refactor all writes to emails to use a default or issuer based bucket

Why keep storage.forceIssuerEmails up to date for all addresses?
`user.checkAuthenticationAndSync` is called without us knowing which specific 
email the user is trying to forceIssuer on.

### CODE REVIEW

File: bin/keysigner [hunk 1]
```diff
@@ -79,11 +80,18 @@ try {
 
 
 // and our single function
-app.post('/wsapi/cert_key', validate({ 'email': 'email', 'pubkey': 'pubkey', 'ephemeral': 'boolean' }), function(req, resp) {
+app.post('/wsapi/cert_key', validate(certKey.args), function(req, resp) {
   var startTime = new Date();
+  var forceIssuer = req.params.forceIssuer;
+  if ('default' !== forceIssuer &&
```
Comment (seanmonstar) [inline bin/keysigner]: is `forceIssuer` required? It doesn't look like it in `wsapi/cert_key`.

If it's not passed, then the `undefined` value would trigger an error. Is this correct behavior?

Comment (ozten) [inline bin/keysigner]: Yes, it's required.

AFAICT any `args` property on a `/wsapi/*.js` module is required.

One of the _few_ improvements of this patch is to remove the duplicate declaration of cert_key arguments.

File: lib/db/json.js [hunk 2]
```diff
@@ -300,66 +316,84 @@ function getAndDeleteRowForSecret(secret, cb) {
...
+          addIt();
+        }
+      });
     });
   });
 };
 
+function withType(email, cb) {
+  var domain = email.replace(/[^@]*@/, '');
+  primary.checkSupport(domain, function (err, urls) {
```
Comment (seanmonstar) [inline lib/db/json.js]: This creates an outgoing request, right? Slow servers could make this much slower. Not sure if there's an answer to this, though.

File: lib/db/mysql.js [hunk 3]
```diff
@@ -363,20 +383,35 @@ function getAndDeleteRowForSecret(secret, cb) {
     });
 }
 
+function withType(email, cb) {
```
Comment (seanmonstar) [inline lib/db/mysql.js]: This function exists also in `db/json.js`.

File: bin/keysigner (hunk 1 above)
Comment (ozten) [inline bin/keysigner]: Thanks for showing me how wsapi calls can have optional arguments. I'll remove `default` as a value and make this optional throughout the code.

File: lib/db/mysql.js (hunk 3 above)
Comment (ozten) [inline lib/db/mysql.js]: I'll extract into a new file dbutils.js.

**Comment by ozten** [comment]
Closing - re-worked in PR#2854

**Pull request closed without merging**